File: Paldromic Substring/JS/Expand Around Center.js
//  Expand Around Center (More efficient than brute force)
function longestPalindromeExpandAroundCenter(s) {
    // Handle edge cases: empty or single-character strings
    if (!s || s.length <= 1) return s;

    let longest = ""; // Initialize the longest palindrome
    let longestLength=0;

    // Iterate through each character as a potential center
    for (let i = 0; i < s.length; i++) {
        // Expand around the center for odd-length palindromes
        const odd = expandAroundCenter(s, i, i);
        
        // If the new palindrome is longer, update the longest
        if (odd.length > longestLength) {
            longest = odd;
            longestLength = longest.length;
        }

        // Expand around the center for even-length palindromes
        const even = expandAroundCenter(s, i, i + 1);

        // If the new palindrome is longer, update the longest
        if (even.length > longest.length) {
            longest = even;
        }
    }

    return longest; // Return the longest palindrome found
}

  // Helper function to expand around the center
function expandAroundCenter(s, left, right) {
    // Expand while within bounds and characters match
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        left--; // Move left pointer to the left
        right++; // Move right pointer to the right
    }
    
    // Return the palindrome (adjust indices after loop)
    return s.substring(left + 1, right);
}

File: String/Paldromic Substring/Python/Expand Around Center.py
# Method 2: Expand Around Center (More efficient than brute force)
def longest_palindrome_expand_around_center(s):
    # Finds the longest palindromic substring by expanding around the center.
    if not s or len(s) <= 1:  # Handle empty or single-character strings
        return s

    longest = ""  # Initialize the longest palindrome

    for i in range(len(s)):  # Iterate through each character as a potential center
        
        # Odd length palindromes
        odd = expand_around_center(s, i, i)  # Expand around a single character
        if len(odd) > len(longest):  # Check if new palindrome is longer
            longest = odd  # Update longest palindrome

        # Even length palindromes
        even = expand_around_center(s, i, i + 1)  # Expand around two adjacent characters
        if len(even) > len(longest):  # Check if new palindrome is longer
            longest = even  # Update longest palindrome

    return longest

def expand_around_center(s, left, right):
    # Expands around the center indices to find a palindrome.
    while left >= 0 and right < len(s) and s[left] == s[right]:  # While within bounds and characters match
        left -= 1  # Move left pointer
        right += 1  # Move right pointer
        
    return s[left + 1:right]  # Return the palindrome (adjust indices after loop)